### PR TITLE
Generate Managed Finalizers for Reconcilers

### PR DESCRIPTION
**Pull request by n3wscott**
fixes https://github.com/knative/pkg/issues/1065

Implements option 1. 

Example integration: https://github.com/knative/sample-controller/pull/213/

Integrators have the option to use either managed finalizers by implementing `FinalizeKind`:

```go
func (r *Reconciler) FinalizeKind(ctx context.Context, o *v1alpha1.AddressableService) reconciler.Event {
	// TODO: add custom finalization logic here.
	return nil
}
```

or without finalizers:

```go
// no change
```

### CODE REVIEW

File: codegen/cmd/injection-gen/generators/reconciler_reconciler.go [hunk 1]
```diff
@@ -208,23 +216,43 @@ func (r *reconcilerImpl) Reconcile(ctx context.Context, key string) error {
...
-	//	// There was a difference and updateFinalizers said it updated and did not return an error.
-	//	r.Recorder.Eventf(resource, {{.corev1EventTypeNormal|raw}}, "Updated", "Updated %q finalizers", resource.GetName())
-	//}
+	var reconcileEvent {{.reconcilerEvent|raw}}
+	if resource.GetDeletionTimestamp().IsZero() {
+		if _, ok := r.reconciler.(Finalizer); ok {
+			// For finalizing reconcilers, if this resource is not being deleted, mark the finalizer.
+			r.setFinalizer(resource)
+		}
+		// Reconcile this copy of the resource and then write back any status
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: You need to patch in the finalizer before calling ReconcileKind

Comment (n3wscott) [reply codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: Done.

File: codegen/cmd/injection-gen/generators/reconciler_reconciler.go [hunk 2]
```diff
@@ -336,22 +337,62 @@ func (r *reconcilerImpl) updateFinalizers(ctx context.Context, desired *{{.type|
...
 	if err != nil {
-		return desired, false, err
+		return err
 	}
 
-	update, err := r.Client.{{.type|versionedClientset}}().{{.type|apiGroup}}(desired.Namespace).Patch(existing.Name, types.MergePatchType, patch)
-	return update, true, err
+	_, err = r.Client.{{.type|versionedClientset}}().{{.type|apiGroup}}(resource.Namespace).Patch(resource.Name, types.MergePatchType, patch)
+	if err != nil {
+		r.Recorder.Eventf(resource, {{.corev1EventTypeNormal|raw}}, "FinalizerUpdate",
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: Remove trailing whitespace:
```suggestion
		r.Recorder.Eventf(resource, {{.corev1EventTypeNormal|raw}}, "FinalizerUpdate",
```

File: codegen/cmd/injection-gen/generators/reconciler_reconciler.go [hunk 3]
```diff
@@ -336,22 +337,62 @@ func (r *reconcilerImpl) updateFinalizers(ctx context.Context, desired *{{.type|
...
 	}
 
-	update, err := r.Client.{{.type|versionedClientset}}().{{.type|apiGroup}}(desired.Namespace).Patch(existing.Name, types.MergePatchType, patch)
-	return update, true, err
+	_, err = r.Client.{{.type|versionedClientset}}().{{.type|apiGroup}}(resource.Namespace).Patch(resource.Name, types.MergePatchType, patch)
+	if err != nil {
+		r.Recorder.Eventf(resource, {{.corev1EventTypeWarning|raw}}, "FinalizerUpdateFailed",
+			"Failed to update finalizers for %q: %v", resource.Name, err)
+	} else {
+		r.Recorder.Eventf(resource, {{.corev1EventTypeNormal|raw}}, "FinalizerUpdate",
```
Comment (mattmoor) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: Remove trailing whitespace:
```suggestion
		r.Recorder.Eventf(resource, {{.corev1EventTypeNormal|raw}}, "FinalizerUpdate",
```

**Pull request merged**